File: applets/src/org/rr0/ufoathome/view/draw/DrawView.java
package org.rr0.ufoathome.view.draw;

import org.rr0.ufoathome.view.AbstractView;
import org.rr0.ufoathome.view.gui.MessageEditable;

import java.awt.*;
import java.text.MessageFormat;
import java.util.ResourceBundle;

/**
 * A drawable view.
 *
 * @version 0.3
 * @author Jerôme Beau
 */
public class DrawView extends AbstractView {
    public static final Cursor DEFAULT_CURSOR = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR);
    public static final Cursor CROSS_HAIR_CURSOR = Cursor.getPredefinedCursor(Cursor.CROSSHAIR_CURSOR);
    public static final Cursor WIDTH_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.E_RESIZE_CURSOR);
    public static final Cursor HEIGHT_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.N_RESIZE_CURSOR);
    public static final Cursor NORTH_WEST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.NW_RESIZE_CURSOR);
    public static final Cursor SOUTH_WEST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.SW_RESIZE_CURSOR);
    public static final Cursor NORTH_EAST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.NE_RESIZE_CURSOR);
    public static final Cursor SOUTH_EAST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.SE_RESIZE_CURSOR);
    public static final Cursor MOVE_CURSOR = Cursor.getPredefinedCursor(Cursor.MOVE_CURSOR);

    private Image bufferedImage;
    protected Graphics bufferedGraphics;
    private ResourceBundle messageBundle;
    public static final String REMOVE_SELECTION_ACTION_COMMAND = "DeleteSelection";
    public static final String DELETE_SOURCE_ACTION_COMMAND = "DeleteX";

    public void start() {
        Dimension size = getSize();     // Avoid using getWidth()/getHeight() which is not available from Java 1.1 (default applets)
        this.bufferedImage = createImage(size.width, size.height);
        bufferedGraphics = bufferedImage.getGraphics();
    }

    public void paint(Graphics g) {
        g.drawImage(bufferedImage, 0, 0, this);
    }

    public void paintShape(DrawShape ufoShape) {
        ufoShape.paint(bufferedGraphics);
    }

    public void displayBuffered() {
        paint(getGraphics());
    }

    public PopupMenu getShapeMenu(DrawSelection source, DrawSelection drawSelection, int mouseX, int mouseY, MessageEditable editable) {
        PopupMenu menu = new PopupMenu(source.toString());
        add(menu);
        //        MenuItem toFront = new MenuItem ("To front");
        //        menu.add(toFront);
        //        MenuItem toBack = new MenuItem ("To back");
        //        menu.add(toBack);
        MenuItem removeSelectionItem = new MenuItem(messageBundle.getString(REMOVE_SELECTION_ACTION_COMMAND));
        removeSelectionItem.setActionCommand(REMOVE_SELECTION_ACTION_COMMAND);
        menu.add(removeSelectionItem);

        MenuItem removeUFO = new MenuItem(MessageFormat.format(messageBundle.getString(DELETE_SOURCE_ACTION_COMMAND), new String[]{editable.getTitle()}));
        menu.add(removeUFO);
        return menu;
    }

    public void setMessageBundle(ResourceBundle messagesBundle) {
        this.messageBundle = messagesBundle;
    }

    /**
         * Prevents flickering
         *
         * @param g
         */
    public void update(Graphics g) {
        paint(g);
    }

    public Graphics getBufferedGraphics() {
        return bufferedGraphics;
    }
}

File: applets/src/org/rr0/ufoathome/view/draw/DrawView.ts
import {AbstractView} from "../AbstractView";
import {MessageEditable} from "../gui/MessageEditable";
import {DrawSelection} from "./DrawSelection";
import {DrawShape} from "./DrawShape";
import {Image} from "../gui/Image";
import {ResourceBundle} from "../../ResourceBundle";
import {Cursor} from "../gui/Cursor";
import {Dimension} from "../gui/Dimension";
import {Graphics} from "../gui/Graphics";

/**
 * A drawable view.
 */
export class DrawView extends AbstractView {
  public static DEFAULT_CURSOR: Cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR);
  public static CROSS_HAIR_CURSOR: Cursor = Cursor.getPredefinedCursor(Cursor.CROSSHAIR_CURSOR);
  public static WIDTH_RESIZE_CURSOR: Cursor = Cursor.getPredefinedCursor(Cursor.E_RESIZE_CURSOR);
  public static HEIGHT_RESIZE_CURSOR: Cursor = Cursor.getPredefinedCursor(Cursor.N_RESIZE_CURSOR);
  public static NORTH_WEST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.NW_RESIZE_CURSOR);
  public static SOUTH_WEST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.SW_RESIZE_CURSOR);
  public static NORTH_EAST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.NE_RESIZE_CURSOR);
  public static SOUTH_EAST_RESIZE_CURSOR = Cursor.getPredefinedCursor(Cursor.SE_RESIZE_CURSOR);
  public static MOVE_CURSOR = Cursor.getPredefinedCursor(Cursor.MOVE_CURSOR);

  private bufferedImage: Image;
  protected bufferedGraphics: Graphics;
  private messageBundle: ResourceBundle;
  public static REMOVE_SELECTION_ACTION_COMMAND: String = "DeleteSelection";
  public static DELETE_SOURCE_ACTION_COMMAND: String = "DeleteX";

  public void;

  start() {
    const size: Dimension = this.getSize();     // Avoid using getWidth()/getHeight() which is not available from Java 1.1 (default applets)
    this.bufferedImage = this.createImage(size.width, size.height);
    this.bufferedGraphics = this.bufferedImage.getGraphics();
  }

  public paint(g: Graphics): void {
    g.drawImage(this.bufferedImage, 0, 0, this, this.horizonY, this);
  }

  public paintShape(ufoShape: DrawShape): void {
    ufoShape.paint(this.bufferedGraphics);
  }

  public displayBuffered(): void {
    this.paint(this.getGraphics());
  }

  public getShapeMenu(source: DrawSelection, drawSelection: DrawSelection, mouseX: number, mouseY: number, editable: MessageEditable): PopupMenu {
    const menu = new PopupMenu(source.toString());
    this.add(menu);
    //        MenuItem toFront = new MenuItem ("To front");
    //        menu.add(toFront);
    //        MenuItem toBack = new MenuItem ("To back");
    //        menu.add(toBack);
    const removeSelectionItem: MenuItem = new MenuItem(this.messageBundle.getString(this.REMOVE_SELECTION_ACTION_COMMAND));
    removeSelectionItem.setActionCommand(this.REMOVE_SELECTION_ACTION_COMMAND);
    menu.add(removeSelectionItem);

    const removeUFO: MenuItem = new MenuItem(MessageFormat.format(this.messageBundle.getString(this.DELETE_SOURCE_ACTION_COMMAND), [editable.getTitle()]));
    menu.add(removeUFO);
    return menu;
  }

  public setMessageBundle(messagesBundle: ResourceBundle): void {
    this.messageBundle = messagesBundle;
  }

  /**
   * Prevents flickering
   *
   * @param g
   */
  public update(g: Graphics): void {
    this.paint(g);
  }

  public getBufferedGraphics(): Graphics {
    return this.bufferedGraphics;
  }
}
